lib/LWP/MediaTypes.pm

```diff
@@ -40,61 +40,68 @@ my %suffixEncoding = (
 read_media_types();
 
 
-
-sub guess_media_type
-{
-    my($file, $header) = @_;
+sub guess_media_type {
+    my ($file, $header) = @_;
     return undef unless defined $file;
 
     my $fullname;
     if (ref($file)) {
-	# assume URI object
-	$file = $file->path;
-	#XXX should handle non http:, file: or ftp: URIs differently
+
+        if ($file->can('path')) {
+            $file = $file->path;
+        }
+        elsif ($file->can('filename')) {
+            $fullname = $file->filename;
+        }
+        else {
+            $fullname = "" . $file;
```

Comment:
This condition doesn't appear to get exercised by the tests.  It would be helpful to get some test coverage for this new code.

Reply:
Also, if we merge this, we'll need to update the documentation to explain the new behaviour.

Reply:
I've added two fixup commits for this. If you agree with the documentation changes I'll autosquash them and push them again. Thanks for the feedback.